x-pack/apm-server/main.go

```diff
@@ -162,6 +169,19 @@ func getBadgerDB(storageDir string) (*badger.DB, error) {
 	return badgerDB, nil
 }
 
+func getStorage(db *badger.DB, ttl time.Duration, limit int64) *eventstorage.ShardedReadWriter {
+	storageMu.Lock()
+	defer storageMu.Unlock()
+	if storage == nil {
+		eventCodec := eventstorage.JSONCodec{}
+		// TTL and storage limit are not dynamically configurable
```

Comment:
> Can you remind me what the rationale is behind reusing the exact same Storage is?

Storage creates a pool of read writers (badger transactions). There were 2 issues with keeping `Storage` object per `Processor`:
1. Logic behind the beater reload is that it starts a new processor before shutting down the older one. This might lead to 2 processors running simultaneously and thus introducing race conditions for finalizing traces.
2. Badger uses [serializable snapshot isolation](https://dgraph.io/blog/post/badger-txn/) and our storage is sharded by badger transactions (each read writer is basically a new badger transaction). This introduces race between when a transaction is created. If a transaction for the same `traceID` is created before the previous `traceID` is committed, the new transaction will not see the commit.

    <details>
    <summary>An example test case to showcase point 2</summary>
    
    ```golang
    func TestBadgerSSI(t *testing.T) {
	    tempdir, err := os.MkdirTemp("", "samplingtest")
	    require.NoError(t, err)
    
	    opts := badger.DefaultOptions(tempdir)
	    badgerDB, err := badger.Open(opts)
	    require.NoError(t, err)
    
	    // NewProcessor creates a new badger transaction for each ReadWriter
	    tx1 := badgerDB.NewTransaction(true)
	    require.NoError(t, tx1.Set([]byte("testkey"), []byte("testval")))
	    require.NoError(t, tx1.Set([]byte("testkey:1"), []byte("testval")))
    
	    // NewProcessor creates a new badger transaction for each ReadWriter
	    tx2 := badgerDB.NewTransaction(true)
	    // ReadWriter#Flush commits the transaction on Processor#Stop
	    require.NoError(t, tx1.Commit())
	    require.NoError(t, tx2.Set([]byte("testkey:2"), []byte("testval")))
	    require.NoError(t, tx2.Set([]byte("testkey:3"), []byte("testval")))
    
	    iterOpts := badger.DefaultIteratorOptions
	    iterOpts.Prefix = []byte("testkey")
    
	    // `testkey` and `testkey:1` will not be visible to `tx2`
	    keysSet := map[string]bool{
		    "testkey:2": true,
		    "testkey:3": true,
	    }
    
	    // ReadWriter#ReadTraceEvents reads the trace events for a traceID to push to ES
	    //
	    // If we were to get the keys explicitly by calling GET for keys
	    // `testkey` or `testkey:1` the GET will fail with transaction conflict
	    // since badger SSI will track reads but iterating over keys will miss entries
	    iter := tx2.NewIterator(iterOpts)
	    for iter.Rewind(); iter.Valid(); iter.Next() {
		    item := iter.Item()
		    if item.IsDeletedOrExpired() {
			    continue
		    }
		    key := string(item.Key())
		    _, ok := keysSet[key]
    
		    require.True(t, ok, "unexpected key")
		    delete(keysSet, key)
	    }
	    iter.Close()
    
	    assert.Equal(t, 0, len(keysSet))
	    require.NoError(t, tx2.Commit())
    }
    ```
    </details>

> The alternative being for each processor to hold a mutex when it starts, then acquire storage; then flush/close storage when it stops, and release the mutex.

IIUC, this approach can take care of point 1 but point 2 will not get resolved only by this. Maybe we can employ lazy creation of read writers but I think that might introduce more complexity than the current approach.

For handling TTL and storage limit another option would be to handle them in main lifecycle by detecting changes on reload.

Reply:
> IIUC, this approach can take care of point 1 but point 2 will not get resolved only by this. Maybe we can employ lazy creation of read writers but I think that might introduce more complexity than the current approach.

I think maybe I was unclear. Essentially what I'm suggesting is to not start the 2nd processor until the 1st one has exited, though I described it in a rather round about way. If we only ever have one TBS processor running at any time (i.e. by using a mutex), then we can guarantee that there is only one SharedReadWriter active, hence only one Badger transaction active for the same shard, and hence there should not be any write conflicts.

Reply:
Sorry, I am still not able to see the mutex solution with the second case. The issue with the second point comes due to the creation of a new `ShardedReadWriter` before the old one is flushed. CMIIW, a mutex can solve the case of 2 simultaneously active `ShardedReadWriter` (by putting mutexes in `Processor#Run`) but, as per the current reload logic, we will still create a new processor and thus a new `ShardedReadWriter` before we stop, and subsequently flush, the old storage. Are you suggesting somehow safeguarding NewProcessor creation also with mutexes?

Reply:
For posterity: Vishal and I discussed this live, and I had mistakenly been thinking that we lazily created badger transactions. So, what we can do instead is lazily create the Storage & SharedReadWriter in Processor.Run, and block in ProcessBatch until it has been created. We would hold a global mutex/semaphore in Processor.Run to prevent two processors from accessing storage at any one time.

Reply:
@axw I have attempted another approach to avoid implementing a special case for reloading configuration OR explicitly synchronizing processors.

Since TTL and storage limit are not tied directly to the underlying badger transaction, I have exposed them as arguments to global `ShardedReadWriter` (as well as the underlying ReadWriter(s)). The global `ShardedReadWriter` is wrapped in a processor-lifecycle-scoped ReadWriter which uses the global ReadWriter with the configuration as arguments. This allows us to keep badger transactions globally while allowing for the processor lifecycle to take care of configuration updates.

Let me know your thoughts on the approach.

Reply:
Just scanned over the PR quickly - this looks way better than the ideas I had. I'll give it a proper review early next week.